src/credentials/GroupDataProviderImpl.h

```diff
@@ -0,0 +1,145 @@
+    CHIP_ERROR GetKeyset(chip::FabricIndex fabric_index, uint16_t keyset_id, Keyset & keys) override;
+    CHIP_ERROR RemoveKeyset(chip::FabricIndex fabric_index, uint16_t keyset_id) override;
+    KeysetIterator * IterateKeysets(chip::FabricIndex fabric_index) override;
+
+    // Fabrics
+    CHIP_ERROR RemoveFabric(chip::FabricIndex fabric_index) override;
+
+    // General
+    CHIP_ERROR Decrypt(PacketHeader packetHeader, PayloadHeader & payloadHeader, System::PacketBufferHandle && msg) override;
+
+private:
+    class GroupMappingIteratorImpl : public GroupMappingIterator
+    {
+    public:
+        GroupMappingIteratorImpl(GroupDataProviderImpl & provider, chip::FabricIndex fabric, chip::EndpointId endpoint);
+        size_t Count() override;
+        bool Next(GroupMapping & item) override;
+        void Release() override;
+
+    private:
+        GroupDataProviderImpl & mProvider;
+        chip::FabricIndex mFabric  = 0;
+        chip::EndpointId mEndpoint = 0;
+        chip::GroupId mGroup       = kUndefinedGroupId;
+    };
+
+    class AllStatesIterator : public GroupStateIterator
+    {
+    public:
+        AllStatesIterator(GroupDataProviderImpl & provider);
+        size_t Count() override;
+        bool Next(GroupState & item) override;
+        void Release() override;
+
+    private:
+        GroupDataProviderImpl & mProvider;
+        uint16_t mIndex    = 0;
+        size_t mCount      = 0;
+        size_t mTotalCount = 0;
+    };
+
+    class FabricStatesIterator : public GroupStateIterator
+    {
+    public:
+        FabricStatesIterator(GroupDataProviderImpl & provider, chip::FabricIndex fabric_index);
+        size_t Count() override;
+        bool Next(GroupState & item) override;
+        void Release() override;
+
+    private:
+        GroupDataProviderImpl & mProvider;
+        chip::FabricIndex mFabric = 0;
+        uint16_t mIndex           = 0;
+        size_t mCount             = 0;
+        size_t mTotalCount        = 0;
+    };
+
+    class KeysetIteratorImpl : public KeysetIterator
+    {
+    public:
+        KeysetIteratorImpl(GroupDataProviderImpl & provider, chip::FabricIndex fabric_index);
+        size_t Count() override;
+        bool Next(Keyset & item) override;
+        void Release() override;
+
+    private:
+        GroupDataProviderImpl & mProvider;
+        chip::FabricIndex mFabric = 0;
+        uint16_t mNextId          = 0;
+        size_t mCount             = 0;
+        size_t mIndex             = 0;
+    };
+
+    chip::PersistentStorageDelegate & mStorage;
+    bool mInitialized = false;
+    BitMapObjectPool<GroupMappingIteratorImpl, kIteratorsMax> mEndpointIterators;
+    BitMapObjectPool<AllStatesIterator, kIteratorsMax> mAllStatesIterators;
+    BitMapObjectPool<FabricStatesIterator, kIteratorsMax> mFabricStatesIterators;
+    BitMapObjectPool<KeysetIteratorImpl, kIteratorsMax> mKeysetIterators;
```

Comment:
Please directly include the header that defines BitMapObjectPool. It seems to be an implicit dependency.

Reply:
Included.